File: frontend/src/api/onedrive.ts
import { apiFetch } from './client'

/** Configuración de conexión con OneDrive (HU-05). */
export interface ConexionOneDrive {
  configurado: boolean
  tenant_id: string | null
  client_id: string | null
  /** El backend nunca devuelve el secreto en claro, solo esta versión enmascarada. */
  client_secret_enmascarado: string | null
  secreto_legible: boolean
  drive_id: string | null
  carpeta_cv: string | null
  ultima_validacion: string | null
  documentos_detectados: number | null
}

export interface ConexionPayload {
  tenant_id: string
  client_id: string
  /** Opcional: si se omite, el backend reutiliza el secreto ya almacenado. */
  client_secret?: string
  drive_id: string
  carpeta_cv: string
}

export interface PruebaConexion {
  exito: boolean
  documentos_detectados: number
  mensaje: string
  detalle: string | null
}

export interface GuardarConexion {
  guardado: boolean
  prueba: PruebaConexion
  configuracion: ConexionOneDrive | null
}

export async function obtenerConexion(): Promise<ConexionOneDrive> {
  return apiFetch<ConexionOneDrive>('/configuracion/onedrive')
}

/** Valida sin guardar. */
export async function probarConexion(datos: ConexionPayload): Promise<PruebaConexion> {
  return apiFetch<PruebaConexion>('/configuracion/onedrive/probar', {
    method: 'POST',
    body: JSON.stringify(datos),
  })
}

/** Revalida la conexión ya guardada y actualiza el conteo de documentos. */
export async function probarConexionGuardada(): Promise<PruebaConexion> {
  return apiFetch<PruebaConexion>('/configuracion/onedrive/probar-guardada', { method: 'POST' })
}

/** Guarda solo si la conexión es válida; si no, la anterior queda intacta. */
export async function guardarConexion(datos: ConexionPayload): Promise<GuardarConexion> {
  return apiFetch<GuardarConexion>('/configuracion/onedrive', {
    method: 'PUT',
    body: JSON.stringify(datos),
  })
}

File: backend/schemas/onedrive.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class ConexionOneDriveRequest(BaseModel):
    """
    Datos de conexión (HU-05).

    `client_secret` es opcional: si se omite, se reutiliza el ya almacenado.
    Esto permite corregir la ruta sin volver a escribir la credencial.
    """

    tenant_id: str = Field(min_length=1, max_length=100)
    client_id: str = Field(min_length=1, max_length=100)
    client_secret: Optional[str] = Field(default=None, max_length=400)
    drive_id: str = Field(min_length=1, max_length=200)
    carpeta_cv: str = Field(default="/HojasDeVida", max_length=400)


class ConexionOneDriveResponse(BaseModel):
    """
    Configuración actual. El secreto siempre viaja enmascarado (RNF10): la API
    nunca devuelve la credencial en claro.
    """

    configurado: bool = False
    tenant_id: Optional[str] = None
    client_id: Optional[str] = None
    client_secret_enmascarado: Optional[str] = None
    secreto_legible: bool = True
    drive_id: Optional[str] = None
    carpeta_cv: Optional[str] = None
    ultima_validacion: Optional[datetime] = None
    documentos_detectados: Optional[int] = None


class PruebaConexionResponse(BaseModel):
    """Resultado de "Probar conexión"."""

    exito: bool
    documentos_detectados: int = 0
    mensaje: str
    detalle: Optional[str] = None


class GuardarConexionResponse(BaseModel):
    guardado: bool
    prueba: PruebaConexionResponse
    configuracion: Optional[ConexionOneDriveResponse] = None
